include/nbl/builtin/glsl/ext/Scan/scan.glsl

```diff
@@ -0,0 +1,122 @@
+
+#include <nbl/builtin/glsl/workgroup/arithmetic.glsl>
+#include <nbl/builtin/glsl/math/typeless_arithmetic.glsl>
+#include <nbl/builtin/glsl/limits/numeric.glsl>
+
+#include "nbl/builtin/glsl/ext/Scan/parameters.glsl"
+
+#ifndef _NBL_GLSL_EXT_SCAN_GET_PADDED_DATA_DECLARED_
+
+uint nbl_glsl_ext_Scan_getPaddedData(in uint idx, in uint pad_val, bool is_upsweep);
+int nbl_glsl_ext_Scan_getPaddedData(in uint idx, in int pad_val, bool is_upsweep);
+float nbl_glsl_ext_Scan_getPaddedData(in uint idx, in float pad_val, bool is_upsweep);
+
+#define _NBL_GLSL_EXT_SCAN_GET_PADDED_DATA_DECLARED_
+#endif
+
+#ifndef _NBL_GLSL_EXT_SCAN_GET_PARAMETERS_DEFINED_
+#error "You need to define `nbl_glsl_ext_Scan_getParameters` and mark `_NBL_GLSL_EXT_SCAN_GET_PARAMETERS_DEFINED_`!"
+#endif
+
+#ifndef _NBL_GLSL_EXT_SCAN_GET_PADDED_DATA_DEFINED_
+#error "You need to define `nbl_glsl_ext_Scan_getPaddedData` and mark `_NBL_GLSL_EXT_SCAN_GET_PADDED_DATA_DEFINED_`!"
+#endif
+
+//
+// Upsweep
+//
+
+#define NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(BIN_OP_NAME, TYPE, IDENTITY)\
+void nbl_glsl_ext_Scan_upsweep##BIN_OP_NAME(in uint idx, out TYPE val)\
+{\
+	TYPE data = nbl_glsl_ext_Scan_getPaddedData(idx, IDENTITY, true);\
+	val = (gl_NumWorkGroups.x == 1u) ? nbl_glsl_workgroupExclusive##BIN_OP_NAME(data) : nbl_glsl_workgroupInclusive##BIN_OP_NAME(data);\
+}
+
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(And, uint, ~0u)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(And, int, int(~0u))
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(And, float, uintBitsToFloat(~0u))
+
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Xor, uint, 0u)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Xor, int, 0)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Xor, float, uintBitsToFloat(0u))
+
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Or, uint, 0u)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Or, int, 0)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Or, float, uintBitsToFloat(0u))
+
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Add, uint, 0u)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Add, int, 0)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Add, float, 0.0)
+
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Mul, uint, 1u)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Mul, int, 1)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Mul, float, 1.0)
+
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Min, uint, ~0u)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Min, int, INT_MAX)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Min, float, FLT_INF)
+
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Max, uint, 0u)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Max, int, INT_MIN)
+NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(Max, float, -FLT_INF)
+
+//
+//	Downsweep
+//
+
+shared uint global_offset_uint;
+shared int global_offset_int;
+shared float global_offset_float;
+
+#define NBL_GLSL_EXT_SCAN_DEFINE_DOWNSWEEP(BIN_OP_NAME, TYPE, IDENTITY, BIN_OP, CONV, INVCONV)\
+void nbl_glsl_ext_Scan_downsweep##BIN_OP_NAME(in uint idx, out TYPE val)\
+{\
+	if (gl_LocalInvocationIndex == (_NBL_GLSL_WORKGROUP_SIZE_ - 1u))\
+		global_offset_##TYPE = nbl_glsl_ext_Scan_getPaddedData(idx, IDENTITY, false);\
+	barrier();\
+	TYPE data = global_offset_##TYPE;\
```

Comment:
replace all 4 lines with
```glsl
TYPE data;
const uint last_ix = _NBL_GLSL_WORKGROUP_SIZE_-1u;
if (gl_LocalInvocationIndex==last_ix)
   data = nbl_glsl_ext_Scan_getPaddedData(idx, IDENTITY, false);
data = nbl_glsl_workgroupBroadcast(data,last_ix);
```

Reply:
you could shoot down this TODO at the same time
https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/include/nbl/builtin/glsl/workgroup/ballot.glsl#L141

Reply:
neat